### ISSUE TITLE
Solvers do not respect original assumptions on variables

### ISSUE THREAD

**Issue by warnerjon12** [sympy/sympy#17840]
The way that some (maybe most or all) of the solvers are implemented causes original assumptions on variables to be ignored or violated in some cases. For instance, when solving an inequality, the variables are automatically replaced with dummy variables that are extended real, and some simplifications are applied to the solution under this assumption regardless of whether or not the original variable(s) were extended real. See also issue #5702. Solve also may return complex solutions to equations containing only real variables. See issue #12525.

**Comment by oscarbenjamin** [comment]
I think `solve` needs a general cleanup/reorganisation. Currently there are many ad-hoc pieces that are not combined as systematically as I would like.

**Comment by warnerjon12** [comment]
I would definitely agree with that, after spending some time in the debugger with that code. Reorganization would make it easier to step through when issues do arise.

**Comment by warnerjon12** [comment]
I might work on this, actually. It will take a bit of thought as to how to effectively reorganize this, but I will see. Maybe dispatching based on equation type could be done by defining `_eval_solve` for different expression types representing the left side of an equation (set to zero) or an inequality. That would be much cleaner than what is currently done. In any case, I will look into it

**Comment by oscarbenjamin** [comment]
Another issue to consider is where exactly checksol gets called. At the moment checksol doesn't necessarily see the original equation.

**Comment by oscarbenjamin** [comment]
See also: https://github.com/sympy/sympy/issues/17418#issuecomment-522116780

**Comment by warnerjon12** [comment]
Yeah, debugging for this is going to be painful, but hopefully it should be fairly helpful.

**Comment by warnerjon12** [comment]
Related to this, should we be changing flags back when we change them in solve?

**Comment by oscarbenjamin** [comment]
I guess you mean the parts where the input flag arguments are manipulated midway through the function. Ideally things would organised so that it just isn't necessary to modify those flags at all. Probably every time that happens there should really be a call to another function instead.

**Comment by smichr** [comment]
> In any case, I will look into it

Your energies might better be directed towards `solveset` which is an attempt to make it do just as you have suggested. Or else go back in history (about 9 years) to the time when `solve` was less capable but organized as you are suggesting.
